architecture_records/001-trainer-controller-framework.md

```diff
@@ -58,18 +58,20 @@ We have implemented a trainer callback (see [here](https://huggingface.co/docs/t
 3. Control operation and action that needs to be performed if rule is evaluated to true.
```

Comment:
Can you please separate the proposed changes into a different pull request, and add the following?

In the PR description for the ADR:
- A quick explanation for why this is safer than `eval`, e.g., malicious operations that can be added through `eval`, but not `simpleeval` to motivate why we should potentially add a new dependency to support this

And in the ADR:
- what things like "smaller numbers" mean to prevent DOS problems - I assume these limits are implemented in `simpleeval` and not specifically meant to be handled in this codebase, as this is a problem that is applicable to anything being executed by that library - is that correct?
- How "big numbers" etc that are not executed because of potential DOS attacks bubble up as errors for users executing this library, as it would be nice to keep error messages interpretable where possible

Reply:
@alex-jw-brooks Will create an ADR with more details but for a quick reference you can look at the unit tests that contain some of the malicious strings. It also has examples of the "big numbers".

On safety, mainly 2 points:
- At a high level `simpleeval` is safer than `eval` because it is a whitelist. It simply doesn't have to support the entirety of Python language featues like `eval` has to.
- `simpleeval` also adds runtime checks (for array multiplication, attribute access, big numbers, etc.) to mitigate the impact of malicious code that can cause DOS.